src/session-pool.js

```diff
@@ -435,252 +366,370 @@ SessionPool.prototype.requestStream = function(config) {
-    self.release(session);
-    end(callback);
-  };
+  if (this.isOpen && this.size() < this.options.min) {
+    this.createSessionInBackground_(session.type);
+  }
 
-  return transaction;
+  return this.requestQueue_
+    .add(function() {
+      return session.delete();
+    })
+    .catch(function(err) {
+      self.emit('error', err);
+    });
 };
 
 /**
- * Create a session, then begin a new transaction.
+ * Deletes idle sessions that exceed the maxIdle configuration.
  *
  * @private
  *
- * @returns {Promise} - Resolves to {Session}.
+ * @returns {Promise}
  */
-SessionPool.prototype.createWriteSession_ = function() {
+SessionPool.prototype.evictIdleSessions_ = function() {
   var self = this;
+  var evicted = [];
 
-  return this.createSession_().then(function(session) {
-    if (session instanceof Error) {
-      return Promise.resolve(session);
-    }
+  var maxIdle = this.options.maxIdle;
+  var min = this.options.min;
 
-    var transaction = self.createTransaction_(session);
+  var idle = this.getIdleSessions_();
+  var count = idle.length;
 
-    session.isWriteSession_ = true;
-    session.transaction_ = transaction;
+  while (count-- > maxIdle && count >= min) {
+    evicted.push(idle.pop());
+  }
 
-    return transaction.begin().then(function() {
-      return session;
-    });
+  var requests = evicted.map(function(session) {
+    return self.destroySession_(session);
   });
+
+  return Promise.all(requests);
 };
 
 /**
- * Attempt to retrieve the next available session. If a write request is made,
- * but no write sessions are available, it will get a read session and begin a
- * new transaction.
+ * Retrieves a list of all the idle sessions.
  *
  * @private
  *
- * @param {object=} options - Session type options.
- * @param {boolean} options.write - Whether or not the requested session is a
- *     write session.
- * @param {function} callback - The callback function.
+ * @returns {Session[]}
  */
-SessionPool.prototype.getNextAvailableSession_ = function(options, callback) {
-  var self = this;
-
-  if (is.fn(options)) {
-    callback = options;
-    options = {};
-  }
+SessionPool.prototype.getIdleSessions_ = function() {
+  var idlesAfter = this.options.idlesAfter * 60000;
 
-  // Write sessions can be used for reads and writes.
-  if (this.writePool && this.writePool.free) {
-    this.getWriteSession(callback);
-    return;
-  }
+  return this.available_.filter(function(session) {
+    return Date.now() - session.lastUsed >= idlesAfter;
+  });
+};
 
-  // Use a read session by calling begin transaction on it.
-  if (options.write && this.pool.free) {
-    this.getSession(function(err, session) {
-      if (err) {
-        callback(err);
-        return;
-      }
+/**
+ * Attempts to get a session of a specific type. If the type is unavailable it
+ * may try to use a different type.
+ *
+ * @private
+ *
+ * @param {string} type The desired session type.
+ * @returns {Promise<Session>}
```

Comment:
if we are never able to create a session, will this promise ever be resolved?

Reply:
We wrap any requests in a `race_` function which essentially races against the pool closing and a timeout occurring. So the only way it would not resolve is if the pool never closed, the user chose to opt out of timing out and a session never became available.